src/architecture/web_api.md

```diff
@@ -0,0 +1,57 @@
+   - All of this can be changed later, so simply use your best judgement at this stage. 
+   - Add methods for [construction](https://github.com/servo/servo/blob/4a5ff01e060293721d10289ec56dbd4c58a0969e/components/script/dom/mod.rs#L91)(not to be confused with a `Constructor` that is part of the Web API). 
+   - [Example result](https://github.com/servo/servo/pull/34844/commits/7d73370b0c41a1b00f4b25b7e1b8bf9b67430708).
+
+### Part 2: Writing a First Draft
+1. For each methods of the bindings trait referred to at 3 in Part 1:
+   - In general, follow the structure of the spec: if a method calls into another named algorithm, implement that named algorithm as a separate private method of your struct, that the trait methods calls into. If you later realize this private method can be used from other structs, make it `pub(crate)`. 
+   - For each algorithm step in the spec:
+       - Copy the prose from the spec, but ommit the step number(these change too often). 
+       - Implement the spec in code(which may take more than one line, and might require additional commenting).
+2. Note: there are certain things that are often needed to perform operation as part of an algorithm: 
+   - `SafeJSContext`: can be obtained using `GlobalScope::get_cx()`.
+   - `GlobalScope`: can be obtained using `self.global()` on a `dom_struct`, or `GlobalScope::from_safe_context`
+   - `InRealm`: can be obtained as an argument to the generated trait method, using [this configuration file](https://github.com/servo/servo/blob/4a5ff01e060293721d10289ec56dbd4c58a0969e/components/script_bindings/codegen/Bindings.conf)
+   - `CanGc`: same as above
+   -  It is best to access them as early as possible, say at the top of the trait method implementation, and to pass them down(as ref for `GlobalScope`) as arguments, in the order described above(with any other needed argument coming in between `&GlobalScope` and `InRealm`). 
+2. This should give you a complete first draft. 
+
+### Part 3: running tests and fixing bugs
+1. Now comes the time to identify which [WPT test](https://book.servo.org/hacking/testing.html) to run against your first draft. You can find them [here](https://github.com/servo/servo/tree/168f7ead152c679ba1e0b8cdddd89e66433b512b/tests/wpt).
+    - This may require turning them on, using [this config](https://github.com/servo/servo/blob/168f7ead152c679ba1e0b8cdddd89e66433b512b/tests/wpt/include.ini).
+ 2. Test may fail because:
+     - There is a bug in the code. These should be fixed. 
+     - The test uses other APIs that aren't supported yet(usually `ERROR`). 
+3. Bugs should be fixed. Copilot is of little help here. 
+4. Expected failures can be marked as such, using the process described [here](https://book.servo.org/hacking/testing.html#updating-web-test-expectations).
+5. This part is done when there are no unexpected test results left. 
```

Comment:
Passing 100% of tests is not a realistic expectation for any kind of non-trivial API. I'm noting this down because Part 4 seems to imply that these steps should be done within a single PR.

Reply:
> **unexpected** test results

(emphasis on unexpected)

You should have a clear idea what part of the test you are expected to fail in your PR and why, and document that in a follow-up issue.